pkg/networkservice/common/updatepath/common.go

```diff
@@ -17,74 +17,77 @@
-		conn.Path = &networkservice.Path{}
+	if conn.GetPath() == nil || len(conn.GetPath().GetPathSegments()) == 0 {
+		if conn.Path == nil {
+			conn.Path = &networkservice.Path{}
+		}
+		conn.Path.Index = 0
+		// Replace path for first segment.
+		conn.Path.PathSegments = []*networkservice.PathSegment{
+			{
+				Name: segmentName,
+				Id:   conn.Id,
+			},
+		}
+		return conn, nil
 	}
 
 	path := conn.GetPath()
 
-	// Make sure index isn't out of bound
-	if (len(path.GetPathSegments()) > 0) && int(path.GetIndex()) >= len(path.GetPathSegments()) {
-		return errors.Errorf("NetworkServiceRequest.Connection.Path.Index(%d) >= len(NetworkServiceRequest.Connection.Path.PathSegments)(%d)",
-			path.GetIndex(), len(path.GetPathSegments()))
+	if int(path.Index) < len(path.PathSegments) && path.PathSegments[path.Index].Name == segmentName {
+		// We already in current segment, just update connection Id, no need to increment
+		conn.Id = path.PathSegments[path.Index].Id
+		return conn, nil
 	}
 
-	// If this isn't already one of our tokens for *this* connectionId ... then update the next one
-	if (len(path.GetPathSegments()) > 0) && !(path.GetPathSegments()[path.GetIndex()].GetName() == u.name &&
-		path.GetPathSegments()[path.GetIndex()].GetId() == conn.GetId()) {
-		path.Index++
-	}
+	// We need to move to next item
+	nextIndex := int(path.Index) + 1
 
-	// If we don't already have a PathSegment at the new index... append an empty one
-	if int(path.GetIndex()) >= len(path.GetPathSegments()) {
-		path.PathSegments = append(path.PathSegments, &networkservice.PathSegment{})
+	if nextIndex > len(path.GetPathSegments()) {
+		// We have index > segments count
+		return nil, errors.Errorf("NetworkServiceRequest.Connection.Path.Index+1==%d should be less or equal len(NetworkServiceRequest.Connection.Path.PathSegement)==%d",
+			nextIndex, len(path.GetPathSegments()))
 	}
-
-	// Extract the authInfo:
-	var authInfo credentials.AuthInfo
-	if p, exists := peer.FromContext(ctx); exists {
-		authInfo = p.AuthInfo
-	}
-
-	// Generate the tok
-	tok, expireTime, err := u.tokenGenerator(authInfo)
-	if err != nil {
-		return errors.WithStack(err)
+	if nextIndex < len(path.GetPathSegments()) && path.GetPathSegments()[nextIndex].Name != segmentName {
+		// We have next, but name is different
+		return nil, errors.Errorf("NetworkServiceRequest.Connection.Path.PathSegement[%d].Name should be %s, but it is %s",
+			nextIndex, segmentName, path.GetPathSegments()[nextIndex].Name)
 	}
 
-	// Convert the expireTime to proto
-	expires, err := ptypes.TimestampProto(expireTime)
-	if err != nil {
-		return errors.WithStack(err)
+	if nextIndex >= len(path.GetPathSegments()) {
+		// Generate new Id for Path segment and connection Id.
+		conn.Id = uuid.New().String()
+		path.PathSegments = append(path.PathSegments, &networkservice.PathSegment{
+			Name: segmentName,
+			Id:   conn.Id,
+		})
+	} else {
+		// Just update ID from next segment.
+		conn.Id = path.GetPathSegments()[nextIndex].Id
```

Comment:
I don't quite follow the logic here, could you explain?

Reply:
This is for case we came second time, in this case current Index are one below, so we need to increment it and set conenction.Id. I've updated code to be more clean.